mlir_graphblas/src/lib/GraphBLAS/GraphBLASOps.cpp

```diff
@@ -4,9 +4,247 @@
 #include "GraphBLAS/GraphBLASDialect.h"
 #include "mlir/IR/OpImplementation.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/None.h"
+
+using namespace mlir;
+using namespace mlir::graphblas;
+
+//===--------------------------------------------------------------------===//
+// Helpers
+//===--------------------------------------------------------------------===//
+
+static llvm::Optional<std::string> checkCSROrCSCTensor(Type inputType, int inputIndex) {
+  /* 
+     Negative values for inputIndex indicate that the input type is the return type.
+     Otherwise, inputIndex indicates which arg inputType corresponds to.
+     
+     Returns llvm::None if the given tensor is valid. 
+     Returns a string explaining the problem otherwise.
+   */
+
+  std::string inputName = inputIndex < 0 ? "Return value" : "Operand #"+std::to_string(inputIndex);
+  
+  mlir::sparse_tensor::SparseTensorEncodingAttr sparseEncoding =
+    mlir::sparse_tensor::getSparseTensorEncoding(inputType);
+  if (!sparseEncoding)
+    return inputName+" must be a sparse tensor.";
+  
+  RankedTensorType inputTensorType = inputType.dyn_cast<RankedTensorType>();
+  if (inputTensorType.getRank() != 2)
+    return inputName+" must have rank 2.";
+  
+  ArrayRef<mlir::sparse_tensor::SparseTensorEncodingAttr::DimLevelType> compression =
+    sparseEncoding.getDimLevelType();
+  if (compression[0] != mlir::sparse_tensor::SparseTensorEncodingAttr::DimLevelType::Dense ||
+      compression[1] != mlir::sparse_tensor::SparseTensorEncodingAttr::DimLevelType::Compressed)
+    return inputName+" must have CSR or CSC compression, i.e. must have "
+      "dimLevelType = [ \"dense\", \"compressed\" ] in the sparse encoding.";
+
+  return llvm::None;
+}
+
+//===--------------------------------------------------------------------===//
+// GraphBLAS Ops Methods
+//===--------------------------------------------------------------------===//
+
+static LogicalResult verify(MatrixApplyOp op) {
+  Type inputType = op.input().getType();
+  Type thunkType = op.thunk().getType();
+  Type resultType = op.getResult().getType();
+
+  llvm::Optional<std::string> inputCompressionErrorMessage = checkCSROrCSCTensor(inputType, 0);
+  if (inputCompressionErrorMessage)
+    return op.emitError(inputCompressionErrorMessage.getValue());
+  
+  llvm::Optional<std::string> resultCompressionErrorMessage = checkCSROrCSCTensor(resultType, -1);
+  if (resultCompressionErrorMessage)
+    return op.emitError(resultCompressionErrorMessage.getValue());
+
+  RankedTensorType inputTensorType = inputType.dyn_cast<RankedTensorType>();
+  RankedTensorType resultTensorType = resultType.dyn_cast<RankedTensorType>();
+  
+  if (inputTensorType.getElementType() != thunkType)
+    return op.emitError("Element type of input tensor does not match type of thunk.");
+
+  if (resultTensorType.getElementType() != thunkType)
+    return op.emitError("Element type of result tensor does not match type of thunk.");
```

Comment:
That's good to know. I left a TODO comment in the code in https://github.com/metagraph-dev/mlir-graphblas/commit/ce09940908b210b2b010b1af43bba3c4c8314e74 so that we don't forget.